src/sinks/file.rs

```diff
@@ -0,0 +1,563 @@
+        let mut closing: Vec<EmbeddedFileSink> = self.closing.drain(..).collect();
+        closing.append(&mut recently_outdated);
+
+        for file in closing.into_iter() {
+            self.poll_close(file);
+        }
+    }
+
+    fn poll_close(&mut self, mut file: EmbeddedFileSink) {
+        match file.close() {
+            Err(err) => error!("Error while closing FileSink: {:?}", err),
+            Ok(Async::Ready(())) => debug!("a FileSink closed"),
+            Ok(Async::NotReady) => self.closing.push(file),
+        }
+    }
+
+    fn opened_files(&self) -> Vec<Arc<Path>> {
+        self.partitions.keys().map(|path| path.clone()).collect()
+    }
+}
+
+impl Sink for PartitionedFileSink {
+    type SinkItem = Event;
+    type SinkError = ();
+
+    fn start_send(&mut self, event: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        match self.path.render(&event) {
+            Ok(bytes) => match std::str::from_utf8(&bytes) {
+                Err(err) => {
+                    warn!(
+                        message = "Path produced is not valid UTF-8. Dropping event.",
+                        ?err
+                    );
+                    Ok(AsyncSink::Ready)
+                }
+
+                Ok(path) => {
+                    let path: Arc<Path> = Arc::from(PathBuf::from(path));
+
+                    if self.close_timeout_secs.is_some() {
+                        self.last_accessed.insert(path.clone(), Instant::now());
+                    }
+
+                    let ref mut partition = match self.partitions.entry(path) {
+                        Entry::Occupied(entry) => entry.into_mut(),
+                        Entry::Vacant(entry) => {
+                            let path = entry.key();
+                            let encoding = self.encoding.clone();
+                            let sink = internal::FileSink::new_with_encoding(&path, encoding);
+                            entry.insert(sink)
+                        }
+                    };
+
+                    partition.start_send(event)
+                }
+            },
+
+            Err(missing_keys) => {
+                warn!(
+                    message = "Keys do not exist on the event. Dropping event.",
+                    ?missing_keys
+                );
+                Ok(AsyncSink::Ready)
+            }
+        }
+    }
+
+    fn poll_complete(&mut self) -> Result<Async<()>, Self::SinkError> {
+        self.partitions
+            .iter_mut()
+            .for_each(|(path, partition)| match partition.poll_complete() {
+                Ok(_) => {}
+                Err(()) => error!("Error in downstream FileSink with path {:?}", path),
+            });
+
+        self.collect_old_files();
+
+        debug!(message = "keeping opened", files = ?self.opened_files());
+
+        Ok(Async::Ready(()))
+    }
+}
+
+mod internal {
```

Comment:
I think this should be moved into a `src/sinks/file/file.rs` and the PartitionedSink stuff can go into a mod.rs in that folder.

Reply:
good idea